imxrt1060-hal/src/gpio.rs

```diff
@@ -212,6 +219,84 @@ where
     }
 }
 
+/// GPIO input interrupt configurations.
+///
+/// These configurations do not take effect until
+/// GPIO input interrupts are enabled.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u32)]
+pub enum InterruptConfiguration {
+    // TODO isn't it a bit repetitive to have "Sensitive" on all here?
+    LowLevelSensitive = 0,
+    HighLevelSensitive = 1,
+    RisingEdgeSensitive = 2,
+    FallingEdgeSensitive = 3,
+    EitherEdgeSensitive = 4,
+}
+
+impl<P> GPIO<P, Input>
+where
+    P: Pin,
+{
+    /// Enable (`true`) or disable (`false`) interrupts for this GPIO input.
+    pub fn set_interrupt_enable(&mut self, enable: bool) {
+        cortex_m::interrupt::free(|_| unsafe {
+            ral::modify_reg!(ral::gpio, self.register_block(), IMR, |imr| if enable {
+                imr | self.mask()
+            } else {
+                imr & !self.mask()
+            })
+        });
+    }
+
+    /// Indicates if interrupts are (`true`) or are not (`false`) enabled for this GPIO input.
+    pub fn is_interrupt_enabled(&self) -> bool {
+        unsafe { ral::read_reg!(ral::gpio, self.register_block(), IMR) & self.mask() != 0u32 }
+    }
+
+    /// Set the interrupt configuration for this GPIO input.
+    pub fn set_interrupt_configuration(&mut self, interrupt_configuration: InterruptConfiguration) {
+        if InterruptConfiguration::EitherEdgeSensitive == interrupt_configuration {
+            unsafe {
+                ral::modify_reg!(ral::gpio, self.register_block(), EDGE_SEL, |edge_sel| {
+                    edge_sel | self.mask()
+                });
+            }
+        } else {
+            unsafe {
+                ral::modify_reg!(ral::gpio, self.register_block(), EDGE_SEL, |edge_sel| {
+                    edge_sel & !self.mask()
+                });
```

Comment:
We'll need critical sections around these `modify_reg!` macros. It's due to a design limitation in today's GPIO driver. Since the `EDGE_SEL` register is shared across all GPIO[N] pins, there's a chance for a RMW to be preempted, which could result in an inconsistent state.

The critical sections aren't the best, nor the perfect defense, but I think we need a breaking change and driver re-design to correct it.